ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -80,12 +80,39 @@ private[platform] class MutableCacheBackedContractStore(
       .flatMap {
         case (cached, toBeFetched) if toBeFetched.isEmpty =>
           Future.successful(Some(cached.max))
+
         case (cached, toBeFetched) =>
-          contractsReader
-            .lookupMaximumLedgerTime(toBeFetched)
-            .map(_.map(m => (cached + m).max))
+          readThroughMaximumLedgerTime(toBeFetched.toList, cached.maxOption)
       }
 
+  private def readThroughMaximumLedgerTime(
+      missing: List[ContractId],
+      acc: Option[Timestamp],
+  ): Future[Option[Timestamp]] =
+    missing match {
+      case contractId :: restOfMissing =>
+        readThroughContractsCache(contractId).flatMap {
+          case active: Active =>
+            val newMaximumLedgerTime = Some(
+              (active.createLedgerEffectiveTime :: acc.toList).max
+            )
+            readThroughMaximumLedgerTime(restOfMissing, newMaximumLedgerTime)
+
+          case _: Archived =>
+            Future.failed(MissingContracts(Set(contractId)))
+
+          case NotFound =>
+            // If cannot be found: no create or archive event for the contract.
+            // Since this contract is part of the input, it was able to be looked up once.
+            // So this is the case of a divulged contract, which was not archived.
+            // Divulged contract does not change maximumLedgerTime
+            readThroughMaximumLedgerTime(restOfMissing, acc)
+        }
+
+      case _ =>
```

Comment:
```suggestion
      case [] =>
```
Scala vs. Haskell question. In my Haskell programs I considered it a good safety measure to make pattern matches complete where possible to avoid surprises in future refactorings when new constructors or pattern guards are introduced. Is there a good reason to not apply this rule in Scala as well, and avoid unexpected default cases hiding required changes?

Reply:
sure thing :+1: that would be a Nil here